apm-agent-plugins/apm-spring-resttemplate-plugin/src/main/java/co/elastic/apm/agent/resttemplate/AsyncRestTemplateInstrumentation.java

```diff
@@ -0,0 +1,108 @@
+import net.bytebuddy.matcher.ElementMatcher;
+import org.springframework.http.HttpRequest;
+import org.springframework.http.client.AsyncClientHttpRequest;
+import org.springframework.util.concurrent.ListenableFuture;
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+public class AsyncRestTemplateInstrumentation extends AbstractRestTemplateInstrumentation {
+
+    @Nullable
+    @VisibleForAdvice
+    public static HelperClassManager<RestTemplateInstrumentationHelper<ListenableFuture<?>>> helperClassManager;
+
+    public AsyncRestTemplateInstrumentation(ElasticApmTracer tracer) {
+        super(tracer);
+        synchronized (AsyncRestTemplateInstrumentation.class) {
+            if (helperClassManager == null) {
+                helperClassManager = HelperClassManager.ForAnyClassLoader.of(tracer,
+                    "co.elastic.apm.agent.resttemplate.helper.RestTemplateInstrumentationHelperImpl",
+                    "co.elastic.apm.agent.resttemplate.helper.ListenableFutureWrapper",
+                    "co.elastic.apm.agent.resttemplate.helper.RestTemplateInstrumentationHelperImpl$ListenableFutureWrapperAllocator");
+            }
+        }
+    }
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    private static void beforeExecute(@Advice.This AsyncClientHttpRequest request,
+                                      @Advice.Local("span") Span span) {
+        if (tracer == null || tracer.getActive() == null) {
+            return;
+        }
+        final TraceContextHolder<?> parent = tracer.getActive();
+        span = HttpClientHelper.startHttpClientSpan(parent, Objects.toString(request.getMethod()), request.getURI(),
+            request.getURI().getHost());
+        if (span != null) {
+            span.activate();
+            if (headerSetterHelperManager != null) {
+                TextHeaderSetter<HttpRequest> headerSetter = headerSetterHelperManager.getForClassLoaderOfClass(HttpRequest.class);
+                if (headerSetter != null) {
+                    span.getTraceContext().setOutgoingTraceContextHeaders(request, headerSetter);
+                }
+            }
+        }
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+    public static void afterExecute(@Advice.Local("span") @Nullable Span span,
+                                      @Advice.Return(readOnly = false) @Nullable ListenableFuture<?> listenableFuture,
+                                      @Advice.Thrown @Nullable Throwable t) {
+        if (span != null) {
+            RestTemplateInstrumentationHelper<ListenableFuture<?>> asyncHelper =
+                helperClassManager.getForClassLoaderOfClass(ListenableFuture.class);
+            if (asyncHelper != null && listenableFuture != null) {
+                listenableFuture = asyncHelper.wrapListenableFuture(listenableFuture, span);
```

Comment:
Using a wrapper is problematic in this case. Consider this scenario:

```java
ListenableFuture<ResponseEntity<Void>> future = asyncRestTemplate.exchange(...);
future.get();
// the future is recycled now but we still have a reference to it.
future.isDone(); // erronously returns false
```

It's better to associate the listener to the span via a `WeakConcurrentMap`. See `co.elastic.apm.agent.asynchttpclient.AbstractAsyncHttpClientInstrumentation` for an example.

Reply:
Get it, Thank you very much.